osu.Game.Tests/NonVisual/FramedReplayInputHandlerTest.cs

```diff
@@ -278,6 +280,38 @@ public void TestMultipleFramesSameTime()
             setTime(-100, -100);
         }
 
+        [Test]
+        public void TestReplayFrameSortStability()
+        {
+            const double repeating_time = 5000;
+
+            int data = 0;
+
+            // 1. add a range of frames in which some of them have the constant time 5000, all without any "data".
+            // 2. randomize the frames.
+            // 3. assign "data" to each frame in ascending order.
+            replay.Frames.AddRange(Enumerable.Range(1, 250).Select(i =>
+            {
+                if (RNG.NextBool())
+                    return new TestReplayFrame(repeating_time, true);
+                else
+                    return new TestReplayFrame(i * 1000, true);
+            }).OrderBy(_ => RNG.Next()).Select(f => new TestReplayFrame(f.Time, true, ++data)));
```

Comment:
I'm very not sold on this. The RNG is unseeded, so this test case is essentially non-deterministic. If it ever fails, nobody will ever know why because the seed used will be the system clock value at the time of initialising RNG, and good luck recovering that. And yes, if you seed it then it is possible that an unstable sort, purely by chance, doesn't shuffle the order and the test will be a false positive. But (1) the more same-time frames you have, the lesser the possibility, semi-theoretically speaking, and (2) I'd probably rather risk that than have a test that is undebuggable because you can never literally get the same values again.

Heck, I'd probably even hardcode a simple 4-frame example for this. This isn't theoretically exhaustively testable anyways. The data space is too large.

Reply:
Ah, now I realize how bad this test case I wrote is. I've tried hardcoding a 4-frame example for this but got a false positive (testing before change), I'll try adding more and see.

Reply:
Got it failing on 20-frames, should be good enough now rather than whatever bullshit that was here.